File: src/gestorAplicacion/administracion/Horario.java
package gestorAplicacion.administracion;

import java.io.Serializable;
import java.util.ArrayList;

public class Horario implements Serializable{
    private Grupo[][] horario = new Grupo[7][24];
    private ArrayList<Grupo> grupoContenidos = new ArrayList<Grupo>();
    private static final long serialVersionUID = 1L;
    private static ArrayList<Horario> horariosTotales = new ArrayList<Horario>();

    public Horario() {
        Horario.horariosTotales.add(this);
    }

    public Horario(int diaSemana, int horaInicio, int horaFinal, Grupo grupo) {
        grupoContenidos.add(grupo);
        for (int hora = horaInicio; hora < horaFinal; hora++){
            this.horario[diaSemana][hora] = grupo;
        }
    }

    public Horario(ArrayList<String> horario, Grupo grupo) {
        //Las clases del horario se dan como string en formato dia-horaInicio-horaFinal
        grupoContenidos.add(grupo);
        for (int i = 0;i<horario.size();i++){
            
            String clase = horario.get(i);
            int dia = Integer.parseInt(clase.substring(0, 1))-1;
            int horaInicio = Integer.parseInt(clase.substring(2, 4));
            int horaFinal = Integer.parseInt(clase.substring(5, 7));
            
            for (int hora = horaInicio; hora < horaFinal; hora++){
                this.horario[dia][hora] = grupo;
            }

        }
    }

    public void ocuparHorario(ArrayList<String> horario,  Grupo grupo) {
        
        grupoContenidos.add(grupo);

        for (int i = 0;i<horario.size();i++){
            
            String clase = horario.get(i);
            int dia = Integer.parseInt(clase.substring(0, 1))-1;
            int horaInicio = Integer.parseInt(clase.substring(2, 4));
            int horaFinal = Integer.parseInt(clase.substring(5, 7));
            
            for (int hora = horaInicio; hora < horaFinal; hora++){
                this.horario[dia][hora] = grupo;
            }
        }
    }

    public void ocuparHorario(Grupo grupo) {

        /*
         * Metodo que solo necesita del grupo para agregarlo al horario correspondiente
         */

        grupoContenidos.add(grupo);

        ArrayList<String> horario = grupo.getHorario();
        for (int i = 0;i<horario.size();i++){
            
            String clase = horario.get(i);
            int dia = Integer.parseInt(clase.substring(0, 1))-1;
            int horaInicio = Integer.parseInt(clase.substring(2, 4));
            int horaFinal = Integer.parseInt(clase.substring(5, 7));
            
            for (int hora = horaInicio; hora < horaFinal; hora++){
                this.horario[dia][hora] = grupo;
            }
        }
    }
    
    public void liberarHorario(ArrayList<String> horario) {
        
        for (int i = 0;i<horario.size();i++){
            
            String clase = horario.get(i);
            int dia = Integer.parseInt(clase.substring(0, 1))-1;
            int horaInicio = Integer.parseInt(clase.substring(2, 4));
            int horaFinal = Integer.parseInt(clase.substring(5, 7));
            
            Grupo grupoEliminado=null;

            for (int hora = horaInicio; hora < horaFinal; hora++){
                grupoEliminado = this.horario[dia][hora];
                this.horario[dia][hora] = null;
            }
            grupoContenidos.remove(grupoEliminado);
        }
    }

    public void vaciarHorario(ArrayList<Grupo> grupos) {
        
        for (Grupo grupo : grupos){
            liberarHorario(grupo.getHorario());
        }
    }
    
    public boolean comprobarDisponibilidad(String clase){

        int dia = Integer.parseInt(clase.substring(0, 1))-1;
        int horaInicio = Integer.parseInt(clase.substring(2, 4));
        int horaFinal = Integer.parseInt(clase.substring(5, 7));

        for (int hora = horaInicio; hora < horaFinal; hora++){
            if (this.horario[dia][hora] != null){
                return false;
            }
        }

        return true;
    }


    public boolean comprobarDisponibilidad(ArrayList<String> clases){
        /*
         * Comprobamos la disponibilidad de un conjunto de clases de un grupo
         */
        boolean ok = true;

        for (String i:clases){
            if (!comprobarDisponibilidad(i)){
                ok = false;
            }
        }

        return ok;
        
    }


    public String mostrarHorario(){

        enum DiaSemana {
    LUNES(0, 5), MARTES(1, 6), MIERCOLES(2, 9), JUEVES(3, 6), VIERNES(4, 7), SABADO(5, 6), DOMINGO(6, 7);

    private final int indice;
    private int length;

    DiaSemana(int indice, int length) {
        this.indice = indice;
        this.length = length;
    }

    public static DiaSemana getDiaPorIndice(int indice) {
        for (DiaSemana dia : DiaSemana.values()) {
            if (dia.indice == indice) {
                return dia;
            }
        }
        return null;
    }
}

        String horario = "HORA        LUNES        MARTES        MIERCOLES        JUEVES        VIERNES        SABADO        DOMINGO\n";

        for (int i = 0; i < 24; i++){

            if (i < 9){
                String horaConCeroDelante = "0" + i;
                String horaSiguienteConCeroDelante = "0" + (i+1);
                horario += "" + horaConCeroDelante + "-" + horaSiguienteConCeroDelante + "       ";
            }
            else if (i == 9){
                horario += "09-10" + "       ";
            }
            else{
                horario += "" + i + "-" + (i + 1) + "       ";
            }

            for (int j = 0; j < 7; j++){
                String materia;
                if (this.horario[j][i] == null){
                    materia = "";
                }
                else{
                    materia = this.horario[j][i].getMateria().getAbreviatura();
                }
                int cantidad_espacios = ((DiaSemana.getDiaPorIndice(j).length + 8) - materia.length());
                String espacios = new String(new char[cantidad_espacios]).replace("\0", " ");
                horario += materia + espacios;
            }
            horario += "\n";
        }
        return horario;
    }

    public ArrayList<Grupo> getGrupoContenidos() {
        return grupoContenidos;
    }
    public void setGrupoContenidos(ArrayList<Grupo> grupoContenidos) {
        this.grupoContenidos = grupoContenidos;
    }
    public static ArrayList<Horario> getHorariosTotales() {
        return horariosTotales;
    }
    
}

File: Python/src/gestorAplicacion/administracion/Horario.py
# Autores:
#   EFRAÍN GÓMEZ RAMÍREZ
#   LIBARDO JOSÉ NAVARRO PEDROZO
#   SEBASTIÁN OCAMPO GALVIS
#   MATEO ÁLVAREZ MURILLO

from enum import Enum


class DiaSemana(Enum):
    # Indice y lenght
    LUNES = (0, 5)
    MARTES = (1, 6)
    MIERCOLES = (2, 9)
    JUEVES = (3, 6)
    VIERNES = (4, 7)
    SABADO = (5, 6)
    DOMINGO = (6, 7)

    @classmethod
    def getDiaPorIndice(cls, indice) -> str:
        for dia in cls:
            if dia.value[0] == indice:
                return dia.name
        return None


class Horario:
    _horariosTotales = []

    def __init__(self, diaSemana=0, horaInicio=0, horaFinal=0, grupo=""):
        self._horario = [[None] * 24 for _ in range(7)]
        self._grupoContenidos = []
        if grupo != "":
            self._grupoContenidos.append(grupo)
            for hora in range(horaInicio, horaFinal):
                self._horario[diaSemana][hora] = grupo

    # Por si alguien utiliza los demas constructores que estaban en Java

    # def __init__(self):
    #     Horario._horariosTotales.append(self)
    #     self._horario = [[None] * 24 for _ in range(7)]
    #     self._grupoContenidos = []

    # def __init__(self, horario, grupo):
    #     self._horario = [[None] * 24 for _ in range(7)]
    #     self._grupoContenidos = []
    #     self._grupoContenidos.append(grupo)
    #     for clase in horario:
    #         dia = int(clase[0]) - 1
    #         horaInicio = int(clase[2:4])
    #         horaFinal = int(clase[5:7])
    #         for hora in range(horaInicio, horaFinal):
    #             self._horario[dia][hora] = grupo

    def ocuparHorario(self, grupo, horario=None) -> None:
        if horario == None:
            horario = grupo.getHorario()
        self._grupoContenidos.append(grupo)
        for clase in horario:
            dia = int(clase[0]) - 1
            horaInicio = int(clase[2:4])
            horaFinal = int(clase[5:7])
            for hora in range(horaInicio, horaFinal):
                self._horario[dia][hora] = grupo

    def liberarHorario(self, horario) -> None:
        for clase in horario:
            dia = int(clase[0]) - 1
            horaInicio = int(clase[2:4])
            horaFinal = int(clase[5:7])
            for hora in range(horaInicio, horaFinal):
                grupoEliminado = self._horario[dia][hora]
                self._horario[dia][hora] = None
      
            for pGrupo in self._grupoContenidos:
                if not isinstance(pGrupo,str):
                    if pGrupo.getNumero() == grupoEliminado.getNumero():
                        if pGrupo.getMateria().getNombre()==grupoEliminado.getMateria().getNombre():
                            self._grupoContenidos.remove(pGrupo)    

    def vaciarHorario(self, grupos) -> None:
        for grupo in grupos:
            self.liberarHorario(grupo.getHorario())

    def comprobarDisponibilidadUna(self, clase) -> bool:
        # print("esto es: "+str(clase))
        # clase = clase[0]
        dia = int(clase[0])-1
        horaInicio = int(clase[2:4])
        horaFinal = int(clase[5:7])
        for hora in range(horaInicio, horaFinal):
            if self._horario[dia][hora] is not None:
                return False
        return True

    def comprobarDisponibilidad(self, clases) -> bool:
        for clase in clases:
            if not self.comprobarDisponibilidadUna(clase):
                return False
        return True

    def mostrarHorario(self) -> str:
        horario = "HORA        LUNES        MARTES        MIERCOLES        JUEVES        VIERNES        SABADO        DOMINGO\n"

        for i in range(24):
            if i < 9:
                horaConCeroDelante = "0" + str(i)
                horaSiguienteConCeroDelante = "0" + str(i + 1)
                horario += (
                    horaConCeroDelante + "-" + horaSiguienteConCeroDelante + "       "
                )
            elif i == 9:
                horario += "09-10" + "       "
            else:
                horario += str(i) + "-" + str(i + 1) + "       "

            for j in range(7):
                materia = ""
                if self._horario[j][i] is not None:
                    materia = self._horario[j][i].getMateria().getAbreviatura()
                cantidad_espacios = (len(DiaSemana.getDiaPorIndice(j)) + 8) - len(
                    materia
                )
                espacios = " " * cantidad_espacios
                horario += materia + espacios
            horario += "\n"
        return horario
    
    def mostrarHorario2(self) -> str:
        # Si el dos, y este si es bueno, porque tiene mano de efrain >:O 
        horario = "HORA        LUNES        MARTES        MIERCOLES        JUEVES        VIERNES        SABADO        DOMINGO\n"

        for i in range(6,24):
            if i < 9:
                horaConCeroDelante = "0" + str(i)
                horaSiguienteConCeroDelante = "0" + str(i + 1)
                horario += (
                    horaConCeroDelante + "-" + horaSiguienteConCeroDelante + "       "
                )
            elif i == 9:
                horario += "09-10" + "       "
            else:
                horario += str(i) + "-" + str(i + 1) + "       "

            for j in range(7):
                materia = ""
                if self._horario[j][i] is not None:
                    materia = self._horario[j][i].getMateria().getAbreviatura()
                if j != 6:
                    cantidad_espacios = (len(DiaSemana.getDiaPorIndice(j)) + 8) - len(materia)
                else:
                    cantidad_espacios = len(materia)
                espacios = " " * cantidad_espacios
                horario += materia + espacios
            horario += "\n"
        return horario

    def getGrupoContenidos(self):
        return self._grupoContenidos

    def setGrupoContenidos(self, grupoContenidos):
        self._grupoContenidos = grupoContenidos

    @classmethod
    def getHorariosTotales(cls):
        return cls._horariosTotales

    @classmethod
    def setHorariosTotales(cls, horarios):
        cls._horariosTotales = horarios
